File: VeilySocialProfile/Windows/src/profile_codec.cpp
#include "profile_codec.h"
#include <cstring>
#include <fstream>
#include <sstream>
#include <stdexcept>

namespace vs {
namespace {

class Writer {
public:
    std::vector<uint8_t> data;
    void u8(uint8_t v) { data.push_back(v); }
    void b(bool v) { u8(v ? 1 : 0); }
    void u16(uint16_t v) { u8(static_cast<uint8_t>(v)); u8(static_cast<uint8_t>(v >> 8)); }
    void u32(uint32_t v) { for (int i=0;i<4;i++) u8(static_cast<uint8_t>(v >> (8*i))); }
    void i32(int32_t v) { u32(static_cast<uint32_t>(v)); }
    void f32(float v) { uint32_t n; std::memcpy(&n, &v, 4); u32(n); }
    void str(const std::string& s) {
        if (s.size() > MAX_STRING_BYTES) throw std::runtime_error("string exceeds VSPF limit");
        u32(static_cast<uint32_t>(s.size()));
        data.insert(data.end(), s.begin(), s.end());
    }
};

class Reader {
public:
    const std::vector<uint8_t>& data;
    size_t pos = 0;
    explicit Reader(const std::vector<uint8_t>& d): data(d) {}
    void need(size_t n) { if (pos + n > data.size()) throw std::runtime_error("truncated VSPF payload"); }
    uint8_t u8(){ need(1); return data[pos++]; }
    bool b(){ auto v=u8(); if(v>1) throw std::runtime_error("invalid bool"); return v!=0; }
    uint16_t u16(){ uint16_t v=u8(); v|=static_cast<uint16_t>(u8())<<8; return v; }
    uint32_t u32(){ uint32_t v=0; for(int i=0;i<4;i++) v|=static_cast<uint32_t>(u8())<<(8*i); return v; }
    int32_t i32(){ return static_cast<int32_t>(u32()); }
    float f32(){ uint32_t n=u32(); float v; std::memcpy(&v,&n,4); return v; }
    std::string str(){ uint32_t n=u32(); if(n>MAX_STRING_BYTES) throw std::runtime_error("string exceeds VSPF limit"); need(n); std::string s(reinterpret_cast<const char*>(data.data()+pos),n); pos+=n; return s; }
};

void writeDecoration(Writer& w, const DecorationRef& d){
    w.u8(static_cast<uint8_t>(d.kind)); w.str(d.builtinName); w.str(d.packRecordKey); w.u32(d.itemId); w.str(d.contentHash); w.str(d.basedOnBuiltin);
}
DecorationRef readDecoration(Reader& r){
    DecorationRef d; d.kind=static_cast<DecorationKind>(r.u8()); d.builtinName=r.str(); d.packRecordKey=r.str(); d.itemId=r.u32(); d.contentHash=r.str(); d.basedOnBuiltin=r.str(); return d;
}
void writeBackground(Writer& w, const BackgroundSpec& bg){
    w.u8(static_cast<uint8_t>(bg.kind)); w.u32(bg.solidArgb); w.f32(bg.startX); w.f32(bg.startY); w.f32(bg.endX); w.f32(bg.endY);
    if(bg.stops.size()>MAX_GRADIENT_STOPS) throw std::runtime_error("too many gradient stops");
    w.u8(static_cast<uint8_t>(bg.stops.size()));
    for(const auto& s:bg.stops){ w.f32(s.position); w.u32(s.argb); }
}
BackgroundSpec readBackground(Reader& r){
    BackgroundSpec bg; bg.kind=static_cast<BackgroundKind>(r.u8()); bg.solidArgb=r.u32(); bg.startX=r.f32(); bg.startY=r.f32(); bg.endX=r.f32(); bg.endY=r.f32();
    auto n=r.u8(); if(n>MAX_GRADIENT_STOPS) throw std::runtime_error("too many gradient stops"); bg.stops.clear(); for(uint8_t i=0;i<n;i++) bg.stops.push_back({r.f32(),r.u32()}); return bg;
}
void writeRect(Writer& w,const RectSpec& x){ w.f32(x.x);w.f32(x.y);w.f32(x.width);w.f32(x.height);w.i32(x.zIndex);w.b(x.visible); }
RectSpec readRect(Reader& r){ RectSpec x; x.x=r.f32();x.y=r.f32();x.width=r.f32();x.height=r.f32();x.zIndex=r.i32();x.visible=r.b();return x; }

void writeElement(Writer& w, const std::shared_ptr<Element>& e, size_t depth){
    if(!e) throw std::runtime_error("null element");
    if(depth>MAX_DEPTH) throw std::runtime_error("element nesting too deep");
    w.u8(static_cast<uint8_t>(e->type)); w.str(e->id); w.str(e->name); writeRect(w,e->rect);
    switch(e->type){
        case ElementType::Block:
            w.u8(static_cast<uint8_t>(e->layout)); writeBackground(w,e->background); writeDecoration(w,e->border); w.f32(e->borderThickness); w.b(e->clipChildren); w.b(e->scrollChildren); w.u8(static_cast<uint8_t>(e->sticky));
            if(e->children.size()>UINT16_MAX) throw std::runtime_error("too many children");
            w.u16(static_cast<uint16_t>(e->children.size()));
            for(auto& c:e->children) writeElement(w,c,depth+1);
            break;
        case ElementType::Text:
            w.str(e->text); w.str(e->fontId); w.f32(e->fontSize); w.u32(e->textArgb); w.u8(static_cast<uint8_t>(e->textAlign)); w.b(e->bold); w.b(e->italic); w.b(e->underline); break;
        case ElementType::Link:
            w.str(e->label); w.u8(static_cast<uint8_t>(e->targetType)); w.str(e->target); break;
        case ElementType::Button:
            w.str(e->label); writeDecoration(w,e->buttonDecoration); writeBackground(w,e->background); w.str(e->fontId); w.f32(e->fontSize); w.u32(e->textArgb); w.u8(static_cast<uint8_t>(e->textAlign)); w.b(e->bold); w.b(e->italic); w.b(e->underline); w.u8(static_cast<uint8_t>(e->targetType)); w.str(e->target); break;
        case ElementType::Stamp:
            writeDecoration(w,e->stampDecoration); w.f32(e->rotationDegrees); w.f32(e->opacity); w.b(e->flipX); w.b(e->flipY); break;
        case ElementType::Media:
            w.u8(static_cast<uint8_t>(e->mediaKind)); w.str(e->mediaRecordKey); w.str(e->mediaContentHash); w.u32(e->intrinsicWidth); w.u32(e->intrinsicHeight); w.str(e->mediaTitle); w.str(e->mediaDescription); break;
        case ElementType::Widget:
            w.str(e->widgetLabel); w.str(e->widgetRecordKey); w.u32(e->widgetItemId); w.str(e->widgetSourceHash); w.u32(e->widgetDefaultWidth); w.u32(e->widgetDefaultHeight); w.b(e->widgetWarnOnResize); break;
    }
}

std::shared_ptr<Element> readElement(Reader& r,size_t depth,size_t& count,uint16_t version){
    if(depth>MAX_DEPTH) throw std::runtime_error("element nesting too deep");
    if(++count>MAX_ELEMENTS_PER_PAGE) throw std::runtime_error("too many elements");
    auto e=std::make_shared<Element>(); e->type=static_cast<ElementType>(r.u8()); e->id=r.str(); e->name=r.str(); e->rect=readRect(r);
    switch(e->type){
        case ElementType::Block:{ e->layout=static_cast<LayoutMode>(r.u8()); e->background=readBackground(r); e->border=readDecoration(r); e->borderThickness=r.f32(); e->clipChildren=r.b(); e->scrollChildren=r.b(); e->sticky=static_cast<StickyMode>(r.u8()); auto n=r.u16(); e->children.clear(); for(uint16_t i=0;i<n;i++) e->children.push_back(readElement(r,depth+1,count,version)); break; }
        case ElementType::Text: e->text=r.str(); if(version>=2)e->fontId=r.str(); e->fontSize=r.f32(); e->textArgb=r.u32(); e->textAlign=static_cast<TextAlign>(r.u8()); e->bold=r.b(); e->italic=r.b(); if(version>=2)e->underline=r.b(); break;
        case ElementType::Link: e->label=r.str(); e->targetType=static_cast<LinkTargetType>(r.u8()); e->target=r.str(); break;
        case ElementType::Button: e->label=r.str(); e->buttonDecoration=readDecoration(r); if(version>=2){e->background=readBackground(r);e->fontId=r.str();e->fontSize=r.f32();e->textArgb=r.u32();e->textAlign=static_cast<TextAlign>(r.u8());e->bold=r.b();e->italic=r.b();e->underline=r.b();}else{e->background.solidArgb=0xFFECF0F8;e->fontId="Default1";e->fontSize=16.0f;e->textArgb=0xFF111827;e->textAlign=TextAlign::Center;} e->targetType=static_cast<LinkTargetType>(r.u8()); e->target=r.str(); break;
        case ElementType::Stamp: e->stampDecoration=readDecoration(r); e->rotationDegrees=r.f32(); e->opacity=r.f32(); e->flipX=r.b(); e->flipY=r.b(); break;
        case ElementType::Media: e->mediaKind=static_cast<MediaKind>(r.u8()); e->mediaRecordKey=r.str(); e->mediaContentHash=r.str(); e->intrinsicWidth=r.u32(); e->intrinsicHeight=r.u32(); e->mediaTitle=r.str(); e->mediaDescription=r.str(); break;
        case ElementType::Widget: e->widgetLabel=r.str(); e->widgetRecordKey=r.str(); e->widgetItemId=r.u32(); if(version>=3){e->widgetSourceHash=r.str();e->widgetDefaultWidth=r.u32();e->widgetDefaultHeight=r.u32();e->widgetWarnOnResize=r.b();} break;
        default: throw std::runtime_error("unknown VSPF element type");
    }
    return e;
}

size_t countElements(const std::shared_ptr<Element>& e,size_t depth,bool& valid){ if(!e){valid=false;return 0;} if(depth>MAX_DEPTH){valid=false;return 0;} size_t n=1; for(auto&c:e->children)n+=countElements(c,depth+1,valid); return n; }
void validateElement(const std::shared_ptr<Element>& e,size_t depth,ValidationResult& out){
    if(!out.ok)return;
    if(!e){out={false,"null element"};return;}
    if(depth>MAX_DEPTH){out={false,"nesting exceeds limit"};return;}
    const auto& q=e->rect; if(q.x<0||q.y<0||q.width<=0||q.height<=0||q.x>1||q.y>1||q.width>1||q.height>1){out={false,"element rectangle outside normalized range"};return;}
    if(e->background.stops.size()>MAX_GRADIENT_STOPS){out={false,"too many gradient stops"};return;}
    if(e->background.kind==BackgroundKind::LinearGradient && e->background.stops.size()<2){out={false,"gradient needs at least two stops"};return;}
    for(const auto&s:e->background.stops) if(s.position<0||s.position>1){out={false,"gradient stop outside 0..1"};return;}
    if(e->opacity<0||e->opacity>1){out={false,"stamp opacity outside 0..1"};return;}
    for(auto&c:e->children) validateElement(c,depth+1,out);
}

} // namespace

ValidationResult validateProfile(const ProfileDocument& doc){
    if(doc.pages.empty()) return {false,"profile has no pages"};
    if(doc.pages.size()>MAX_PAGES) return {false,"too many pages"};
    bool found=false; for(const auto&p:doc.pages){ if(p.aspectRatio<0.20f||p.aspectRatio>1.20f)return{false,"page aspect ratio outside 0.20..1.20"}; if(p.id==doc.defaultPageId)found=true; bool valid=true; auto count=countElements(p.root,0,valid); if(!valid||count>MAX_ELEMENTS_PER_PAGE)return{false,"page element/depth limit exceeded"}; ValidationResult r;validateElement(p.root,0,r);if(!r.ok)return r; }
    if(!found)return{false,"default page id does not exist"};
    return {};
}

std::vector<uint8_t> encodeProfileBinary(const ProfileDocument& doc){
    auto vr=validateProfile(doc); if(!vr.ok) throw std::runtime_error(vr.message); Writer w; w.data.insert(w.data.end(),{'V','S','P','F'}); w.u16(FORMAT_VERSION); w.str(doc.profileId); w.str(doc.profileName); w.str(doc.defaultPageId); w.u16(static_cast<uint16_t>(doc.pages.size()));
    for(const auto&p:doc.pages){w.str(p.id);w.str(p.name);w.f32(p.aspectRatio);writeElement(w,p.root,0);} return w.data;
}

ProfileDocument decodeProfileBinary(const std::vector<uint8_t>& bytes){
    Reader r(bytes); if(r.u8()!='V'||r.u8()!='S'||r.u8()!='P'||r.u8()!='F')throw std::runtime_error("not a VSPF profile"); auto v=r.u16(); if(v<1||v>FORMAT_VERSION)throw std::runtime_error("unsupported VSPF version"); ProfileDocument d; d.profileId=r.str();d.profileName=r.str();d.defaultPageId=r.str();auto n=r.u16();if(n==0||n>MAX_PAGES)throw std::runtime_error("invalid page count"); for(uint16_t i=0;i<n;i++){Page p;p.id=r.str();p.name=r.str();p.aspectRatio=r.f32();if(p.aspectRatio<0.20f||p.aspectRatio>1.20f)throw std::runtime_error("invalid page aspect ratio");size_t count=0;p.root=readElement(r,0,count,v);d.pages.push_back(std::move(p));} if(r.pos!=bytes.size())throw std::runtime_error("trailing data in VSPF payload"); auto vr=validateProfile(d);if(!vr.ok)throw std::runtime_error(vr.message);return d;
}

static const char* B64="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
std::string base64Encode(const std::vector<uint8_t>& data){std::string out;int val=0,valb=-6;for(uint8_t c:data){val=(val<<8)+c;valb+=8;while(valb>=0){out.push_back(B64[(val>>valb)&0x3F]);valb-=6;}}if(valb>-6)out.push_back(B64[((val<<8)>>(valb+8))&0x3F]);while(out.size()%4)out.push_back('=');return out;}
std::vector<uint8_t> base64Decode(const std::string& text){std::vector<int>T(256,-1);for(int i=0;i<64;i++)T[(unsigned char)B64[i]]=i;std::vector<uint8_t> out;int val=0,valb=-8;for(unsigned char c:text){if(c=='='||c=='\r'||c=='\n'||c==' '||c=='\t')continue;if(T[c]==-1)throw std::runtime_error("invalid base64");val=(val<<6)+T[c];valb+=6;if(valb>=0){out.push_back(static_cast<uint8_t>((val>>valb)&0xFF));valb-=8;}}return out;}
std::string encodeProfileText(const ProfileDocument& doc){return std::string("VEILYSOCIAL_PROFILE_V2\n")+base64Encode(encodeProfileBinary(doc))+"\n";}
ProfileDocument decodeProfileText(const std::string& text){const std::string h1="VEILYSOCIAL_PROFILE_V1",h2="VEILYSOCIAL_PROFILE_V2";auto p=text.find('\n');if(p==std::string::npos||(text.substr(0,p)!=h1&&text.substr(0,p)!=h2))throw std::runtime_error("invalid profile text envelope");return decodeProfileBinary(base64Decode(text.substr(p+1)));}
void saveProfileTextFile(const ProfileDocument& doc,const std::string& path){std::ofstream f(path,std::ios::binary|std::ios::trunc);if(!f)throw std::runtime_error("could not create profile file");auto s=encodeProfileText(doc);f.write(s.data(),static_cast<std::streamsize>(s.size()));if(!f)throw std::runtime_error("could not write profile file");}
ProfileDocument loadProfileTextFile(const std::string& path){std::ifstream f(path,std::ios::binary);if(!f)throw std::runtime_error("could not open profile file");std::ostringstream s;s<<f.rdbuf();return decodeProfileText(s.str());}

} // namespace vs

File: Android/Source/app/src/main/java/app/weave/ProfileCodec.kt
package app.weave

import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.charset.StandardCharsets
import java.util.Base64

object ProfileCodec {
    private const val ENVELOPE = "VEILYSOCIAL_PROFILE_V2"
    private const val LEGACY_ENVELOPE = "VEILYSOCIAL_PROFILE_V1"

    private class W {
        val out = ByteArrayOutputStream()
        fun u8(v: Int) { out.write(v and 0xFF) }
        fun bool(v: Boolean) = u8(if (v) 1 else 0)
        fun u16(v: Int) { u8(v); u8(v ushr 8) }
        fun u32(v: Long) { repeat(4) { u8((v ushr (it * 8)).toInt()) } }
        fun i32(v: Int) = u32(v.toLong() and 0xFFFFFFFFL)
        fun f32(v: Float) = u32(java.lang.Float.floatToRawIntBits(v).toLong() and 0xFFFFFFFFL)
        fun str(s: String) {
            val b = s.toByteArray(StandardCharsets.UTF_8)
            require(b.size <= VspfLimits.MAX_STRING_BYTES) { "string exceeds VSPF limit" }
            u32(b.size.toLong()); out.write(b)
        }
    }

    private class R(private val data: ByteArray) {
        var pos = 0
        private fun need(n: Int) { require(pos + n <= data.size) { "truncated VSPF payload" } }
        fun u8(): Int { need(1); return data[pos++].toInt() and 0xFF }
        fun bool(): Boolean { val v = u8(); require(v <= 1) { "invalid bool" }; return v != 0 }
        fun u16(): Int = u8() or (u8() shl 8)
        fun u32(): Long { var v = 0L; repeat(4) { v = v or (u8().toLong() shl (it * 8)) }; return v }
        fun i32(): Int = u32().toInt()
        fun f32(): Float = java.lang.Float.intBitsToFloat(u32().toInt())
        fun str(): String {
            val n = u32().toInt(); require(n in 0..VspfLimits.MAX_STRING_BYTES) { "string exceeds VSPF limit" }
            need(n); val s = String(data, pos, n, StandardCharsets.UTF_8); pos += n; return s
        }
        fun done() = pos == data.size
    }

    data class Validation(val ok: Boolean, val message: String = "")

    fun validate(doc: ProfileDocument): Validation {
        if (doc.pages.isEmpty()) return Validation(false, "profile has no pages")
        if (doc.pages.size > VspfLimits.MAX_PAGES) return Validation(false, "too many pages")
        if (doc.pages.none { it.id == doc.defaultPageId }) return Validation(false, "default page id does not exist")
        for (p in doc.pages) {
            if (p.aspectRatio !in .20f..1.20f) return Validation(false, "page aspect ratio outside 0.20..1.20")
            var count = 0
            fun walk(e: Element, depth: Int): String? {
                if (depth > VspfLimits.MAX_DEPTH) return "nesting exceeds limit"
                count++; if (count > VspfLimits.MAX_ELEMENTS_PER_PAGE) return "page element limit exceeded"
                val q = e.rect
                if (q.x !in 0f..1f || q.y !in 0f..1f || q.width !in .0001f..1f || q.height !in .0001f..1f) return "element rectangle outside normalized range"
                if (e.background.stops.size > VspfLimits.MAX_GRADIENT_STOPS) return "too many gradient stops"
                if (e.background.kind == BackgroundKind.LinearGradient && e.background.stops.size < 2) return "gradient needs at least two stops"
                if (e.background.stops.any { it.position !in 0f..1f }) return "gradient stop outside 0..1"
                if (e.opacity !in 0f..1f) return "stamp opacity outside 0..1"
                e.children.forEach { child -> walk(child, depth + 1)?.let { return it } }
                return null
            }
            walk(p.root, 0)?.let { return Validation(false, it) }
        }
        return Validation(true)
    }

    private fun writeDecoration(w: W, d: DecorationRef) {
        w.u8(d.kind.ordinal); w.str(d.builtinName); w.str(d.packRecordKey); w.u32(d.itemId); w.str(d.contentHash); w.str(d.basedOnBuiltin)
    }
    private fun readDecoration(r: R) = DecorationRef(
        kind = DecorationKind.entries[r.u8()], builtinName = r.str(), packRecordKey = r.str(), itemId = r.u32(), contentHash = r.str(), basedOnBuiltin = r.str()
    )
    private fun writeBackground(w: W, b: BackgroundSpec) {
        w.u8(b.kind.ordinal); w.u32(b.solidArgb.toLong() and 0xFFFFFFFFL); w.f32(b.startX); w.f32(b.startY); w.f32(b.endX); w.f32(b.endY)
        require(b.stops.size <= VspfLimits.MAX_GRADIENT_STOPS); w.u8(b.stops.size)
        b.stops.forEach { w.f32(it.position); w.u32(it.argb.toLong() and 0xFFFFFFFFL) }
    }
    private fun readBackground(r: R): BackgroundSpec {
        val b = BackgroundSpec(kind = BackgroundKind.entries[r.u8()], solidArgb = r.u32().toInt(), startX = r.f32(), startY = r.f32(), endX = r.f32(), endY = r.f32(), stops = mutableListOf())
        val n = r.u8(); require(n <= VspfLimits.MAX_GRADIENT_STOPS) { "too many gradient stops" }
        repeat(n) { b.stops += GradientStop(r.f32(), r.u32().toInt()) }
        return b
    }
    private fun writeRect(w: W, q: RectSpec) { w.f32(q.x); w.f32(q.y); w.f32(q.width); w.f32(q.height); w.i32(q.zIndex); w.bool(q.visible) }
    private fun readRect(r: R) = RectSpec(r.f32(), r.f32(), r.f32(), r.f32(), r.i32(), r.bool())

    private fun writeElement(w: W, e: Element, depth: Int) {
        require(depth <= VspfLimits.MAX_DEPTH)
        w.u8(e.type.ordinal + 1); w.str(e.id); w.str(e.name); writeRect(w, e.rect)
        when (e.type) {
            ElementType.Block -> {
                w.u8(e.layout.ordinal); writeBackground(w, e.background); writeDecoration(w, e.border); w.f32(e.borderThickness); w.bool(e.clipChildren); w.bool(e.scrollChildren); w.u8(e.sticky.ordinal)
                require(e.children.size <= 65535); w.u16(e.children.size); e.children.forEach { writeElement(w, it, depth + 1) }
            }
            ElementType.Text -> { w.str(e.text); w.str(e.fontId); w.f32(e.fontSize); w.u32(e.textArgb.toLong() and 0xFFFFFFFFL); w.u8(e.textAlign.ordinal); w.bool(e.bold); w.bool(e.italic); w.bool(e.underline) }
            ElementType.Link -> { w.str(e.label); w.u8(e.targetType.ordinal); w.str(e.target) }
            ElementType.Button -> { w.str(e.label); writeDecoration(w, e.buttonDecoration); writeBackground(w, e.background); w.str(e.fontId); w.f32(e.fontSize); w.u32(e.textArgb.toLong() and 0xFFFFFFFFL); w.u8(e.textAlign.ordinal); w.bool(e.bold); w.bool(e.italic); w.bool(e.underline); w.u8(e.targetType.ordinal); w.str(e.target) }
            ElementType.Stamp -> { writeDecoration(w, e.stampDecoration); w.f32(e.rotationDegrees); w.f32(e.opacity); w.bool(e.flipX); w.bool(e.flipY) }
            ElementType.Media -> { w.u8(e.mediaKind.ordinal); w.str(e.mediaRecordKey); w.str(e.mediaContentHash); w.u32(e.intrinsicWidth); w.u32(e.intrinsicHeight); w.str(e.mediaTitle); w.str(e.mediaDescription) }
            ElementType.Widget -> { w.str(e.widgetLabel); w.str(e.widgetRecordKey); w.u32(e.widgetItemId); w.str(e.widgetSourceHash); w.u32(e.widgetDefaultWidth); w.u32(e.widgetDefaultHeight); w.bool(e.widgetWarnOnResize) }
        }
    }

    private fun readElement(r: R, depth: Int, count: IntArray, version: Int): Element {
        require(depth <= VspfLimits.MAX_DEPTH) { "element nesting too deep" }
        count[0]++; require(count[0] <= VspfLimits.MAX_ELEMENTS_PER_PAGE) { "too many elements" }
        val tag = r.u8(); require(tag in 1..7) { "unknown VSPF element type" }
        val e = Element(type = ElementType.entries[tag - 1], id = r.str(), name = r.str(), rect = readRect(r))
        when (e.type) {
            ElementType.Block -> {
                e.layout = LayoutMode.entries[r.u8()]; e.background = readBackground(r); e.border = readDecoration(r); e.borderThickness = r.f32(); e.clipChildren = r.bool(); e.scrollChildren = r.bool(); e.sticky = StickyMode.entries[r.u8()]
                val n = r.u16(); e.children.clear(); repeat(n) { e.children += readElement(r, depth + 1, count, version) }
            }
            ElementType.Text -> {
                e.text = r.str()
                if (version >= 2) e.fontId = r.str()
                e.fontSize = r.f32(); e.textArgb = r.u32().toInt(); e.textAlign = TextAlignMode.entries[r.u8()]; e.bold = r.bool(); e.italic = r.bool()
                if (version >= 2) e.underline = r.bool()
            }
            ElementType.Link -> { e.label = r.str(); e.targetType = LinkTargetType.entries[r.u8()]; e.target = r.str() }
            ElementType.Button -> {
                e.label = r.str(); e.buttonDecoration = readDecoration(r)
                if (version >= 2) {
                    e.background = readBackground(r); e.fontId = r.str(); e.fontSize = r.f32(); e.textArgb = r.u32().toInt(); e.textAlign = TextAlignMode.entries[r.u8()]; e.bold = r.bool(); e.italic = r.bool(); e.underline = r.bool()
                } else {
                    e.background = BackgroundSpec(solidArgb = 0xFFECF0F8.toInt()); e.fontId = "Default1"; e.fontSize = 16f; e.textArgb = 0xFF111827.toInt(); e.textAlign = TextAlignMode.Center
                }
                e.targetType = LinkTargetType.entries[r.u8()]; e.target = r.str()
            }
            ElementType.Stamp -> { e.stampDecoration = readDecoration(r); e.rotationDegrees = r.f32(); e.opacity = r.f32(); e.flipX = r.bool(); e.flipY = r.bool() }
            ElementType.Media -> { e.mediaKind = MediaKind.entries[r.u8()]; e.mediaRecordKey = r.str(); e.mediaContentHash = r.str(); e.intrinsicWidth = r.u32(); e.intrinsicHeight = r.u32(); e.mediaTitle = r.str(); e.mediaDescription = r.str() }
            ElementType.Widget -> { e.widgetLabel = r.str(); e.widgetRecordKey = r.str(); e.widgetItemId = r.u32(); if (version >= 3) { e.widgetSourceHash = r.str(); e.widgetDefaultWidth = r.u32(); e.widgetDefaultHeight = r.u32(); e.widgetWarnOnResize = r.bool() } }
        }
        return e
    }

    fun encodeBinary(doc: ProfileDocument): ByteArray {
        val v = validate(doc); require(v.ok) { v.message }
        val w = W(); "VSPF".forEach { w.u8(it.code) }; w.u16(VspfLimits.FORMAT_VERSION); w.str(doc.profileId); w.str(doc.profileName); w.str(doc.defaultPageId); w.u16(doc.pages.size)
        doc.pages.forEach { p -> w.str(p.id); w.str(p.name); w.f32(p.aspectRatio); writeElement(w, p.root, 0) }
        return w.out.toByteArray()
    }

    fun decodeBinary(data: ByteArray): ProfileDocument {
        val r = R(data); require(r.u8() == 'V'.code && r.u8() == 'S'.code && r.u8() == 'P'.code && r.u8() == 'F'.code) { "not a VSPF profile" }
        val version = r.u16(); require(version in 1..VspfLimits.FORMAT_VERSION) { "unsupported VSPF version" }
        val doc = ProfileDocument(profileId = r.str(), profileName = r.str(), defaultPageId = r.str(), pages = mutableListOf())
        val n = r.u16(); require(n in 1..VspfLimits.MAX_PAGES) { "invalid page count" }
        repeat(n) {
            val id = r.str(); val name = r.str(); val aspect = r.f32(); require(aspect in .20f..1.20f) { "invalid page aspect ratio" }
            doc.pages += Page(id = id, name = name, aspectRatio = aspect, root = readElement(r, 0, intArrayOf(0), version))
        }
        require(r.done()) { "trailing data in VSPF payload" }
        val v = validate(doc); require(v.ok) { v.message }; return doc
    }

    fun encodeText(doc: ProfileDocument): String = ENVELOPE + "\n" + Base64.getEncoder().encodeToString(encodeBinary(doc)) + "\n"
    fun decodeText(text: String): ProfileDocument {
        val first = text.lineSequence().firstOrNull() ?: error("empty profile file")
        require(first == ENVELOPE || first == LEGACY_ENVELOPE) { "invalid profile text envelope" }
        val b64 = text.substringAfter('\n').filterNot { it.isWhitespace() }
        return decodeBinary(Base64.getDecoder().decode(b64))
    }
    fun save(doc: ProfileDocument, file: File) { file.parentFile?.mkdirs(); file.writeText(encodeText(doc), Charsets.UTF_8) }
    fun load(file: File): ProfileDocument = decodeText(file.readText(Charsets.UTF_8))
}
